samples/extensions/sparse_image/sparse_image.cpp

```diff
@@ -0,0 +1,1661 @@
+	std::array<VkPipelineShaderStageCreateInfo, 2> shader_stages{};
+
+	// Vertex stage of the pipeline
+	shader_stages[0] = load_shader("sparse_image/shader.vert", VK_SHADER_STAGE_VERTEX_BIT);
+	shader_stages[1] = load_shader("sparse_image/shader.frag", VK_SHADER_STAGE_FRAGMENT_BIT);
+
+	// We need to specify the pipeline layout and the render pass description up front as well.
+	VkGraphicsPipelineCreateInfo pipeline_create_info = vkb::initializers::pipeline_create_info(sample_pipeline_layout, render_pass);
+	pipeline_create_info.stageCount                   = vkb::to_u32(shader_stages.size());
+	pipeline_create_info.pStages                      = shader_stages.data();
+	pipeline_create_info.pVertexInputState            = &vertex_input;
+	pipeline_create_info.pInputAssemblyState          = &input_assembly;
+	pipeline_create_info.pRasterizationState          = &raster;
+	pipeline_create_info.pColorBlendState             = &blend;
+	pipeline_create_info.pMultisampleState            = &multisample;
+	pipeline_create_info.pViewportState               = &viewport;
+	pipeline_create_info.pDepthStencilState           = &depth_stencil;
+	pipeline_create_info.pDynamicState                = &dynamic;
+
+	VK_CHECK(vkCreateGraphicsPipelines(get_device().get_handle(), pipeline_cache, 1, &pipeline_create_info, nullptr, &sample_pipeline));
+}
+
+void sparse_image::setup_camera()
+{
+	camera.type = vkb::CameraType::FirstPerson;
+	camera.set_perspective(fov_degrees, static_cast<float>(width) / static_cast<float>(height), 0.1f, 512.0f);
+	camera.set_rotation(glm::vec3(0.0f, 0.0f, 0.0f));
+	camera.set_translation(glm::vec3(0.0f, 0.0f, -1.0f));
+	camera.translation_speed = 1.0f;
+}
+
+
+VkCommandBuffer sparse_image::begin_single_time_command()
+{
+	VkCommandBufferAllocateInfo alloc_info{};
+	alloc_info.sType             = VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO;
+	alloc_info.level             = VK_COMMAND_BUFFER_LEVEL_PRIMARY;
+	alloc_info.commandPool       = cmd_pool;
+	alloc_info.commandBufferCount = 1;
+
+	VkCommandBuffer command_buffer;
+	vkAllocateCommandBuffers(device->get_handle(), &alloc_info, &command_buffer);
+
+	VkCommandBufferBeginInfo begin_info{};
+	begin_info.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO;
+	begin_info.flags = VK_COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT;
+
+	vkBeginCommandBuffer(command_buffer, &begin_info);
+
+	return command_buffer;
+}
+
+
+void sparse_image::end_single_time_commands(VkCommandBuffer command_buffer)
+{
+	vkEndCommandBuffer(command_buffer);
+
+	VkSubmitInfo submitInfo{};
+	submitInfo.sType              = VK_STRUCTURE_TYPE_SUBMIT_INFO;
+	submitInfo.commandBufferCount = 1;
+	submitInfo.pCommandBuffers    = &command_buffer;
+
+	vkQueueSubmit(queue, 1, &submitInfo, VK_NULL_HANDLE);
+	vkQueueWaitIdle(queue);
+
+	vkFreeCommandBuffers(device->get_handle(), cmd_pool, 1, &command_buffer);
+}
+
+
+void sparse_image::generate_mips()
+{
+	VkCommandBuffer command_buffer = begin_single_time_command();
```

Comment:
use device->create_command_buffer();
and end it device->flush_command_buffer();